core/src/main/scala/kafka/zk/migration/ZkConfigMigrationClient.scala

```diff
@@ -0,0 +1,250 @@
+import kafka.zookeeper.{CreateRequest, SetDataRequest}
+import org.apache.kafka.common.config.{ConfigDef, ConfigResource}
+import org.apache.kafka.common.errors.InvalidRequestException
+import org.apache.kafka.common.metadata.ClientQuotaRecord.EntityData
+import org.apache.kafka.common.quota.ClientQuotaEntity
+import org.apache.kafka.metadata.migration.{ConfigMigrationClient, MigrationClientException, ZkMigrationLeadershipState}
+import org.apache.zookeeper.KeeperException.Code
+import org.apache.zookeeper.{CreateMode, KeeperException}
+
+import java.{lang, util}
+import java.util.Properties
+import java.util.function.BiConsumer
+import scala.collection.Seq
+import scala.jdk.CollectionConverters._
+
+class ZkConfigMigrationClient(
+  zkClient: KafkaZkClient,
+  passwordEncoder: PasswordEncoder
+) extends ConfigMigrationClient with Logging {
+
+  val adminZkClient = new AdminZkClient(zkClient)
+
+  override def iterateClientQuotas(
+    quotaEntityConsumer: BiConsumer[util.List[EntityData], util.Map[String, lang.Double]]
+  ): Unit = {
+    def migrateEntityType(entityType: String): Unit = {
+      adminZkClient.fetchAllEntityConfigs(entityType).foreach { case (name, props) =>
+        val entity = List(new EntityData().setEntityType(entityType).setEntityName(name)).asJava
+        val quotaMap = ZkAdminManager.clientQuotaPropsToDoubleMap(props.asScala).map {
+          case (key, value) => key -> lang.Double.valueOf(value)
+        }.toMap.asJava
+        quotaEntityConsumer.accept(entity, quotaMap)
+      }
+    }
+
+    migrateEntityType(ConfigType.User)
+    migrateEntityType(ConfigType.Client)
+
+    adminZkClient.fetchAllChildEntityConfigs(ConfigType.User, ConfigType.Client).foreach { case (name, props) =>
+      // Taken from ZkAdminManager
+      val components = name.split("/")
+      if (components.size != 3 || components(1) != "clients")
+        throw new IllegalArgumentException(s"Unexpected config path: ${name}")
+      val entity = List(
+        new EntityData().setEntityType(ConfigType.User).setEntityName(components(0)),
+        new EntityData().setEntityType(ConfigType.Client).setEntityName(components(2))
+      )
+      val quotaMap = props.asScala.map { case (key, value) =>
+        val doubleValue = try lang.Double.valueOf(value) catch {
+          case _: NumberFormatException =>
+            throw new IllegalStateException(s"Unexpected client quota configuration value: $key -> $value")
+        }
+        key -> doubleValue
+      }.asJava
+      quotaEntityConsumer.accept(entity.asJava, quotaMap)
+    }
+
+    migrateEntityType(ConfigType.Ip)
+  }
+
+  override def iterateBrokerConfigs(configConsumer: BiConsumer[String, util.Map[String, String]]): Unit = {
+    val brokerEntities = zkClient.getAllEntitiesWithConfig(ConfigType.Broker)
+    zkClient.getEntitiesConfigs(ConfigType.Broker, brokerEntities.toSet).foreach { case (broker, props) =>
+      val brokerResource = if (broker == ConfigEntityName.Default) {
+        ""
```

Comment:
we should have a comment that the empty string means "default entity name" in kraft

or possibly use a constant, but I don't know if there is one